src/System.Net.Security/tests/FunctionalTests/KerberosTest.cs

```diff
@@ -0,0 +1,488 @@
+            using (Process clearCreds = Process.Start(startInfo))
+            {
+                clearCreds.WaitForExit();
+                output.WriteLine("kdestroy returned {0}", clearCreds.ExitCode);
+                return (clearCreds.ExitCode == 0);
+            }
+        }
+
+        private static int RunSetupScript(string args = null)
+        {
+            ProcessStartInfo startInfo = new ProcessStartInfo();
+
+            // since ProcessStartInfo does not support Verb, we use sudo as
+            // the program to be run
+            startInfo.FileName = SudoCommand;
+            startInfo.Arguments = string.Format("bash {0} {1}", ScriptName, args);
+            using (Process kdcSetup = Process.Start(startInfo))
+            {
+                kdcSetup.WaitForExit();
+                return kdcSetup.ExitCode;
+            }
+        }
+    }
+
+    public class KerberosTest : IDisposable, IClassFixture<KDCSetup>
+    {
+        private readonly byte[] _firstMessage = Encoding.UTF8.GetBytes("Sample First Message");
+        private readonly byte[] _secondMessage = Encoding.UTF8.GetBytes("Sample Second Message");
+        private readonly bool _isKrbAvailable; // tests are no-op if kerberos is not available on the host machine
+        private readonly KDCSetup _fixture;
+        private readonly ITestOutputHelper _output;
+
+        public KerberosTest(KDCSetup fixture, ITestOutputHelper output)
+        {
+            _fixture = fixture;
+            _output = output;
+            _isKrbAvailable = _fixture.CheckAndClearCredentials(_output);
+        }
+
+        [Fact, OuterLoop]
+        [PlatformSpecific(PlatformID.Linux)]
+        public void NegotiateStream_StreamToStream_KerberosAuthentication_Success()
+        {
+            if (!_isKrbAvailable)
+            {
+                _output.WriteLine("skipping NegotiateStream_StreamToStream_KerberosAuthentication_Success");
+                return;
+            }
+
+            VirtualNetwork network = new VirtualNetwork();
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new UnixGssFakeNegotiateStream(serverStream))
+            {
+                Assert.False(client.IsAuthenticated, "client is not authenticated");
+
+                Task[] auth = new Task[2];
+                string user = string.Format("{0}@{1}", TestConfiguration.KerberosUser, TestConfiguration.Realm);
+                string target = string.Format("{0}@{1}", TestConfiguration.HostTarget, TestConfiguration.Realm);
+                NetworkCredential credential = new NetworkCredential(user, _fixture.password);
+                auth[0] = client.AuthenticateAsClientAsync(credential, target);
+                auth[1] = server.AuthenticateAsServerAsync();
```

Comment:
Rather than 3 magic numbers (`new Task[2]`, `auth[0]`, `auth[1]`), consider

```C#
Task[] auth =
{
    client.AuthenticateAsClientAsync(credential, target),
    server.AuthenticateAsServerAsync(),
};
```